pubsublite/internal/wire/assigner.go

```diff
@@ -0,0 +1,169 @@
+	"google.golang.org/grpc"
+
+	vkit "cloud.google.com/go/pubsublite/apiv1"
+	pb "google.golang.org/genproto/googleapis/cloud/pubsublite/v1"
+)
+
+// partitionSet is a set of partition numbers.
+type partitionSet map[int]struct{}
+
+func newPartitionSet(assignmentpb *pb.PartitionAssignment) partitionSet {
+	var void struct{}
+	partitions := make(map[int]struct{})
+	for _, p := range assignmentpb.GetPartitions() {
+		partitions[int(p)] = void
+	}
+	return partitionSet(partitions)
+}
+
+func (ps partitionSet) Ints() (partitions []int) {
+	for p := range ps {
+		partitions = append(partitions, p)
+	}
+	return
+}
+
+func (ps partitionSet) Contains(partition int) bool {
+	_, exists := ps[partition]
+	return exists
+}
+
+// A function that generates a 16-byte UUID.
+type generateUUIDFunc func() (uuid.UUID, error)
+
+// partitionAssignmentReceiver must enact the received partition assignment from
+// the server, or otherwise return an error, which will break the stream. The
+// receiver must not call the assigner, as this would result in a deadlock.
+type partitionAssignmentReceiver func(partitionSet) error
+
+// assigner wraps the partition assignment stream and notifies a receiver when
+// the server sends a new set of partition assignments for a subscriber.
+type assigner struct {
+	// Immutable after creation.
+	assignmentClient  *vkit.PartitionAssignmentClient
+	initialReq        *pb.PartitionAssignmentRequest
+	receiveAssignment partitionAssignmentReceiver
+
+	// Fields below must be guarded with mu.
+	stream *retryableStream
+
+	abstractService
+}
+
+func newAssigner(ctx context.Context, assignmentClient *vkit.PartitionAssignmentClient, genUUID generateUUIDFunc, settings ReceiveSettings, subscriptionPath string, receiver partitionAssignmentReceiver) (*assigner, error) {
+	clientID, err := genUUID()
+	if err != nil {
+		return nil, fmt.Errorf("pubsublite: failed to generate client UUID: %v", err)
+	}
+
+	a := &assigner{
+		assignmentClient: assignmentClient,
+		initialReq: &pb.PartitionAssignmentRequest{
+			Request: &pb.PartitionAssignmentRequest_Initial{
+				Initial: &pb.InitialPartitionAssignmentRequest{
+					Subscription: subscriptionPath,
+					ClientId:     clientID[:],
+				},
+			},
+		},
+		receiveAssignment: receiver,
+	}
+	a.stream = newRetryableStream(ctx, a, settings.Timeout, reflect.TypeOf(pb.PartitionAssignment{}))
+	return a, nil
+}
+
+func (a *assigner) Start() {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if a.unsafeUpdateStatus(serviceStarting, nil) {
+		a.stream.Start()
+	}
+}
+
+func (a *assigner) Stop() {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.unsafeInitiateShutdown(serviceTerminating, nil)
+}
+
+func (a *assigner) newStream(ctx context.Context) (grpc.ClientStream, error) {
+	return a.assignmentClient.AssignPartitions(ctx)
+}
+
+func (a *assigner) initialRequest() (interface{}, bool) {
+	return a.initialReq, false // No initial response expected
+}
+
+func (a *assigner) validateInitialResponse(_ interface{}) error {
```

Comment:
hmm. How did you configure the retryable stream to not have an initial response?

Reply:
The second return value of `initialRequest` is a bool indicating whether an initial response is expected. I will refactor it to a strong type in a separate PR (pulls in other unrelated changes) to make it clearer.

As for why - in long running integration tests, I found that when the partition assignment stream reconnects and the client UUID hasn't changed, the server doesn't bother to return the initial set of partition assignments (so my test would just hang waiting for it, then timeout and fail).